improver/regrid/bilinear.py

```diff
@@ -0,0 +1,502 @@
+        four_mismatch = np.concatenate((four_mismatch, leftover_idw))
+        weights, indexes, surface_type_mask = lakes_islands(
+            four_mismatch,
+            weights,
+            indexes,
+            surface_type_mask,
+            in_latlons,
+            out_latlons,
+            in_classified,
+            out_classified,
+            in_lons_dim,
+            vicinity,
+        )
+
+    return weights, indexes
+
+
+def one_mismatched_input_point(
+    one_mismatch_indexes,
+    surface_type_mask,
+    indexes,
+    weights,
+    out_latlons,
+    in_latlons,
+    in_lons_dim,
+):
+    """
+    updating source points and weighting for one mismatched source-point cases
+
+    Args:
+        one_mismatch_indexes(numpy.ndarray):
+            selected target points which have 1 false source point
+        surface_type_mask(numpy.ndarray)):
+            numpy ndarray of bool, true if source point type matches target point type
+        indexes(numpy.ndarray):
+            array of source grid point number for all target grid points
+        weights(numpy.ndarray):
+            array of source grid point weighting for all target grid points
+        in_latlons(numpy.ndarray):
+            tource points's latitude-longitudes
+        out_latlons(numpy.ndarray):
+            target points's latitude-longitudes
+        in_lons_dim (int):
+            source grid's longitude dimension
+
+    Returns:
+         Tuple[numpy.ndarray,numpy.ndarray]:
+            updated weights: array of source grid point weighting for target grid points
+            excluded_indexes: target points which are not handled in this function
+    """
+
+    # Set up input points spacing values
+    lat_spacing = in_latlons[in_lons_dim, 0] - in_latlons[0, 0]
+    lon_spacing = in_latlons[1, 1] - in_latlons[0, 1]
+    lat_lon_area = lat_spacing * lon_spacing
+    excluded_indexes = np.array([], dtype=int)
+
+    # Process groups of output points which have each one of the four surrounding
+    # input points as mismatched surface type
+    for i in range(4):
+        # Determine group of output points to process in this iteration
+        inverse_mask_i = (
+            np.where(np.logical_not(surface_type_mask[one_mismatch_indexes, i]))
+        )[0]
+        indexes_with_i_mismatched = one_mismatch_indexes[inverse_mask_i]
+
+        # Extract subset of output lat/lon
+        out_lats = out_latlons[indexes_with_i_mismatched, 0]
+        out_lons = out_latlons[indexes_with_i_mismatched, 1]
+
+        # Calculate lat/lon of left bottom input point
+        lat_1 = in_latlons[indexes[indexes_with_i_mismatched, 0], 0]
+        lon_1 = in_latlons[indexes[indexes_with_i_mismatched, 0], 1]
+
+        # Calculate updated weights
+        if i == 0:
+            lat2_lat = lat_1 + lat_spacing - out_lats
+            lon3_lon = lon_1 + lon_spacing - out_lons
+            weight2 = lat_spacing * lon3_lon / lat_lon_area
```

Comment:
It is basically using triangle interpolation function, but we keep four weighting so that we can handle it in unified way for follow-up calculation. All interpolation formulations are explicitly given for 4 different cases. We also deferred cases if the point is not within the triangle.

Reply:
I think this will all be addressed by follow on documentation PR now so resolving thread